exporter/prometheusremotewriteexporter/wal_test.go

```diff
@@ -149,3 +152,57 @@ func TestWAL_persist(t *testing.T) {
 	require.Equal(t, reqLFromWAL[0], reqL[0])
 	require.Equal(t, reqLFromWAL[1], reqL[1])
 }
+
+func TestWAL_E2E(t *testing.T) {
+	in := []*prompb.WriteRequest{
+		series("mem_used_percent", 0, 0),
+		series("mem_used_percent", 15, 34),
+		series("mem_used_percent", 30, 99),
+	}
+	out := make([]*prompb.WriteRequest, 0, len(in))
+
+	done := make(chan struct{})
+	sink := func(_ context.Context, reqs []*prompb.WriteRequest) error {
+		out = append(out, reqs...)
+		if len(out) >= len(in) {
+			close(done)
+		}
+		return nil
+	}
+
+	wal := newWAL(&WALConfig{
+		Directory: t.TempDir(),
+	}, sink)
+	defer os.RemoveAll(t.TempDir())
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = contextWithLogger(ctx, zaptest.NewLogger(t))
+
+	if err := wal.run(ctx); err != nil {
+		require.NoError(t, err)
+	}
+
+	if err := wal.persistToWAL(in); err != nil {
+		require.NoError(t, err)
+	}
+
+	// wait until the tail routine is no longer busy
+	wal.rNotify <- struct{}{}
```

Comment:
Just to confirm: if I remove this, the test will be blocked on the `<-done` until it times out. That's the purpose of this test, and what you are fixing here, right?

Reply:
I'm also a bit confused as to why this is needed here.  `run()` will fork and (indirectly) call `read()`, blocking until it returns, which will happen when a record is read from the WAL or the context is cancelled.  `read()` will try to read from the WAL, find nothing, and then block reading from `rNotify`.  `persistToWAL()` will then acquire the mutex, signal on `rNotify`, and write to the WAL.  Once `persistToWAL()` signals on `rNotify` then `read()` starts blocking on acquiring the mutex in `tryRead()`, eventually acquiring it once `persistToWAL()` executes its deferred unlock.  `read()` will then get a successful result from `tryRead()`, popping all the way back out to `continuallyPopWALThenExport()` which may or may not actually export (probably not with the default truncate frequency and buffer size), and then goes right back into `read()`.  At this point the WAL is empty and `tryRead()` returns the sentinel error `read()` uses to decide to wait on `rNotify`.  Since `persistToWAL()` has returned nothing is going to send anything on `rNotify` until this line here.  Sending a value here causes `read()` to try again, but nothing happens.  Then the context is cancelled and a bunch of `select` statements hit their `<-ctx.Done()` case, causing `read()` to return `nil, context.Canceled` to `continuallyPopWALThenExport()`, which returns `context.Canceled` to `run()`, executing its deferred export along the way, causing `sink` to close `done` and allowing this function to reach its requirement that the exported data equals the input data.

tl;dr: I don't see what writing to this channel does that isn't also accomplished by cancelling the context.

Reply:
@sh0rez , are you still interested in pursuing this?